### ISSUE TITLE
jax.lax.psum hangs when using TPU v3-32 

### ISSUE THREAD

**Issue by anirudhitagi** [google/jax#18224]
### Description

Greetings,

I am attempting to run all-reduce (jax.lax.psum) on TPU v3-32. 
Here is the command I used to create the tpu nodes::

```bash
gcloud compute tpus tpu-vm create node-v3-32-euw4-a --project synergyexp \
      --zone=europe-west4-a \
      --accelerator-type=v3-32 \
      --version=tpu-ubuntu2204-base \
      --preemptible

gcloud compute tpus tpu-vm ssh node-v3-32-euw4-a --zone=europe-west4-a --worker=all --command="pip install jax[tpu] -f https://storage.googleapis.com/jax-releases/libtpu_releases.html"
```

I have been using this python script to do the all reduce:
```python
# jax_allreduce_v3_32.py
# The following code snippet will be run on all TPU hosts
import sys
import jax
import jax.numpy as jnp
from jax import pmap
import timeit
import time

size_of_allreduce_in_bytes = [512]
#size_of_allreduce_in_bytes = [1073741824]
iterations = 1

for ss in size_of_allreduce_in_bytes:

    device_count = jax.device_count()

    x = jnp.arange(ss/(4*device_count))
    x1 = [x,x,x,x,x,x,x,x] # for 8 TPUs
    x2 = jnp.array(x1)

    print(f"running collective of size = {x2.nbytes}")

    # The number of TPU cores attached to this host
    local_device_count = jax.local_device_count()

    # The psum is performed over all mapped devices across the Pod
    # xs = jax.numpy.ones(jax.local_device_count())

    start_time = time.time()
    for i in range(iterations):
        r = jax.pmap(lambda xx: jax.lax.psum(xx, 'i'), axis_name='i')(x2)
    end_time = time.time()



    # Print from a single host to avoid duplicated output
    if jax.process_index() == 0:
        print(f"execution_time = {(end_time-start_time)/iterations}")
        print('global device count:', jax.device_count())
        print('local device count:', jax.local_device_count())
        print('pmap result:', r)
```

```bash
$ gcloud compute tpus tpu-vm ssh node-v3-32-euw4-a \
    --zone=europe-west4-a \
    --worker=all \
    --command="python3 jax_allreduce_v3_32.py"

Using ssh batch size of 1. Attempting to SSH into 1 nodes with a total of 4 workers.
SSH: Attempting to connect to worker 0...
SSH: Attempting to connect to worker 1...
SSH: Attempting to connect to worker 2...
SSH: Attempting to connect to worker 3...
running collective of size = 128
running collective of size = 128
running collective of size = 128


```
and it just hangs there indefinitely. 
Any advice on what is going wrong and how to fix this?

Best regards,
AI

### What jax/jaxlib version are you using?

jax 0.4.19, jaxlib 0.4.19

### Which accelerator(s) are you using?

TPU

### Additional system info

tpu-ubuntu2204-base

**Comment by pschuh** [comment]
Looks like one of the workers failed to even start for some reason.

Your script worked for me (I only modified the print stmt to print out the worker id).

```
Using ssh batch size of 1. Attempting to SSH into 1 nodes with a total of 4 workers.
SSH: Attempting to connect to worker 0...
SSH: Attempting to connect to worker 1...
SSH: Attempting to connect to worker 2...
SSH: Attempting to connect to worker 3...
running collective of size = 128 for worker: 2
running collective of size = 128 for worker: 3
running collective of size = 128 for worker: 1
running collective of size = 128 for worker: 0
execution_time = 0.276763916015625
global device count: 32
local device count: 8
pmap result: [[ 0. 32. 64. 96.]
 [ 0. 32. 64. 96.]
 [ 0. 32. 64. 96.]
 [ 0. 32. 64. 96.]
 [ 0. 32. 64. 96.]
 [ 0. 32. 64. 96.]
 [ 0. 32. 64. 96.]
 [ 0. 32. 64. 96.]]
```

you can also try putting:
print("Initialization complete: %s" % jax.process_index()) before anything else to see if all the processes have initialized properly.

Consider looking through (or attaching) the logs for each process to see if there is some problem initializing:
`gcloud compute tpus tpu-vm ssh node-v3-32-euw4-a     --zone=europe-west4-a     --worker=all     --command="cat /tmp/tpu_logs/tpu_driver.INFO"`

**Comment by anirudhitagi** [comment]
Hi @pschuh,
Thank you for your response!

I tried it again and it did work.
However, when I increased the ``` iterations = 1``` to ``` iterations = 1000```

It hangs and terminates with this message:

```
running collective of size = 128
running collective of size = 128
running collective of size = 128
##### Command execution on worker 0 failed with exit status 1. Continuing.
```

Here is the log I collected as you suggested:
[temp.log](https://github.com/google/jax/files/13179880/temp.log)


any idea why this is happening?
For context, I am running the iteration 1000 times so that I get an averaged time of the collective communication execution. 

Best regards,
AI

**Comment by pschuh** [comment]
Hi, another possibility is that you're early exiting from one of the tasks. Consider fetching the final pmap result on all tasks in order to ensure that all tasks have finished the pmap result.

Also consider lifting the pmap like so:

```
pmapped_fn = jax.pmap(lambda xx: jax.lax.psum(xx, 'i'), axis_name='i')

...

     for i in range(iterations):
        r = pmapped_fn(x2)
```

I'm noticing in your log that there are many recompiles.

**Comment by anirudhitagi** [comment]
Hi @pschuh ,

Thank you so much for the suggestion!
it looks like it is working!

I am new this, so please forgive me if my doubt is too basic:

How is this logic executing:
```python
pmapped_fn = jax.pmap(lambda xx: jax.lax.psum(xx, 'i'), axis_name='i')

for i in range(iterations):
        r = pmapped_fn(x2)
```


is the pmap executing `iterations` times or , is the result being assigned to `r` `iterations` times

I ask because here is the execution times (in s):
1. iteration = 1       --- execution_time = 0.27577686309814453
2. iteration = 100   --- execution_time = 0.003425896167755127
3. iteration = 1000 --- execution_time = 0.000904909372329712

Best Regards,
AI

**Comment by pschuh** [comment]
Because jax uses a jit compilation where pmapped_fn is lazily compiled on the first step, we can expect that the time breakdown is approximately: `jit_time + n_steps * (max(dispatch_overhead, on_device_time))` Because your on-device function is trivial (just a all-reduce of a trivial value), this means that your code is: `jit_time + n_steps * dispatch_overhead`. From your numbers, it looks like the jit time is roughly ~270ms and the dispatch_overhead is ~600us. You can see more with: https://jax.readthedocs.io/en/latest/profiling.html (This will also give you a better estimate of the on_device_time). For performance, you want your entire model step update (fwd + bwd + reduce-gradients + optimizer apply) to be a single training step. I would also recommend switching to the jit API with something like this (This will trigger the SPMD partitioner and make it easier to try different shardings in the future):

```
mesh = jax.sharding.Mesh(jax.devices(), ("data",))
weight_sharding = jax.sharding.NamedSharding(mesh, jax.sharding.PartitionSpec())
data_sharding = jax.sharding.NamedSharding(mesh, jax.sharding.PartitionSpec("data"))

@functools.partial(jax.jit, in_shardings=(weight_sharding, data_sharding, data_sharding), out_shardings=(weight_sharding))
def fake_update_step(weights, x, y):
  def fake_loss_fn(weights, x):
    x = x @ weights
    return ((x - y) ** 2).sum()
  loss, grads = jax.value_and_grad(fake_loss_fn)(weights, x)
  weights -= learning_rate * grads
  return weights, loss
  
weights = init_weights_somehow()
for x, y in dataset:
  weights, _ = fake_update_step(weights, x, y)
```

**Comment by anirudhitagi** [comment]
Thank you so much for the detailed solution!
I will take a look into the suggested options. 

Best regards,
AI

**Comment by anirudhitagi** [comment]
Hi @pschuh,

I had a question regarding - the formula `jit_time + n_steps * dispatch_overhead`

How are you calculating `n_steps`? I believe it depends on the type of all-reduce
I guess my question is what type of all-reduce is performed by `jax.lax.psum`? is it Ring, Binary tree, halving-doubling or something else?
